### PR TITLE
new nlu trainer

### PR DESCRIPTION
**Pull request by tcnguyen**
- Create botfuel-trainer-nlu with new trainer api
- Modify dialog-manager to handle the case where multiple intents detected ==> propose resolution dialog
- Modify previous botfuel-nlu to work with new dialog-manager
- Remove multiple intent test in test-complexdialogs

### CODE REVIEW

File: packages/botfuel-dialog/src/nlus/intent.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const logger = require('logtown')('Intent');
+const SdkError = require('../errors/sdk-error');
+
+/** Intent class */
+class Intent {
+  /**
```
Comment (KevinDepuydt) [inline packages/botfuel-dialog/src/nlus/intent.js]: I Would have define Intent types as static properties (as in the Dialog class) and I would used it in the other classes where this is needed:

```javascript
static TYPE_QNA = 'QnA';
static TYPE_INTENT = 'Intent';

get TYPE_QNA() {
  return Intent.TYPE_QNA;
}

get TYPE_INTENT() {
  return Intent.TYPE_INTENT;
}
```

By doing this you make sure that the value of these types are defined in only one place which is very useful if the value have to change in the future.

Comment (tcnguyen) [reply packages/botfuel-dialog/src/nlus/intent.js]: Thank you. Good point. Will apply this.

Comment (tcnguyen) [reply packages/botfuel-dialog/src/nlus/intent.js]: thank you. good point!

File: packages/botfuel-dialog/src/dialog-manager.js [hunk 2]
```diff
@@ -113,26 +96,42 @@ class DialogManager extends Resolver {
...
+          name: intents[0].name,
+          entities: intents[0].answers,
+        };
+      } else {
+        newDialog = {
+          name: intents[0].name,
+          entities,
+        };
+      }
+    }
```
Comment (windkomo) [inline packages/botfuel-dialog/src/dialog-manager.js]: A ternary here is cool for readability:

```javascript
const firstIntent = intents[0];
newDialog = {
     name: firstIntent.name,
     entities: firstIntent.isQnA() ? firstIntent.answers : entities
};
```

File: packages/botfuel-dialog/src/dialog-manager.js [hunk 3]
```diff
@@ -47,31 +47,14 @@ class DialogManager extends Resolver {
...
+   * Returns the last "reentrant" dialog to execute if no other dialog is found.
+   * When the sentence itself does not contain enough information for the DialogManager
+   * to compute a dialog, the DialogManager recalls the first reentrant dialog from the
+   * stack of previous dialogs
    * @param {Object[]} previousDialogs - the previous dialogs
    * @returns {String} a dialog name
    */
-  getLastDialog(previousDialogs) {
+  getLastReentrantDialog(previousDialogs) {
     for (let i = previousDialogs.length - 1; i >= 0; i--) {
```
[the remark this answers is not in the document]
Comment (tcnguyen) [reply packages/botfuel-dialog/src/dialog-manager.js]: great js! thanks

File: packages/botfuel-dialog/src/config.js [hunk 4]
```diff
@@ -39,6 +39,7 @@ const defaultConfig = {
     intentThreshold: 0.8,
   },
   multiIntent: false,
+  trainerApiUri: 'https://trainer-api-staging.herokuapp.com/api/v0',
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/config.js]: this is temporary, right?

File: packages/botfuel-dialog/src/dialog-manager.js [hunk 5]
```diff
@@ -47,39 +47,20 @@ class DialogManager extends Resolver {
...
-      return 0;
-    });
-  }
-
-  /**
-   * Returns the last dialog to execute if no other dialog is found.
+   * Returns the last "reentrant" dialog to execute if no other dialog is found.
+   * When the sentence itself does not contain enough information for the DialogManager
+   * to compute a dialog, the DialogManager recalls the first reentrant dialog from the
+   * stack of previous dialogs
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/dialog-manager.js]: Missing dot at the end of sentence.

File: packages/botfuel-dialog/src/dialog-manager.js [hunk 6]
```diff
@@ -113,26 +94,35 @@ class DialogManager extends Resolver {
...
-      const name = intents[i];
-      const characteristics = this.resolve(name).characteristics;
-      if (characteristics.reentrant) {
-        nb++;
+
+    let newDialog = null;
+    if (intents.length > 1) {
+      newDialog = {
+        name: 'intent-resolution',
+        entities: {
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/dialog-manager.js]: data: {entities, intents}

File: packages/botfuel-dialog/src/dialogs/intent-resolution-dialog.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * limitations under the License.
+ */
+const logger = require('logtown')('IntentResolutionDialog');
+const Dialog = require('./dialog');
+
+/**
+ * The default resolution dialog when multiple intities are detected from the trainer
+ * @extends IntentResolutionDialog
+ */
+class IntentResolutionDialog extends Dialog {
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/dialogs/intent-resolution-dialog.js]: Rename into ClassificationDisambiguationDialog or something similar

Comment (yangeorget) [inline packages/botfuel-dialog/src/dialogs/intent-resolution-dialog.js]: Should extend TextDialog (and in the future BaseDialog)

File: packages/botfuel-dialog/src/index.js [hunk 8]
```diff
@@ -22,6 +22,7 @@ const BotTextMessage = require('./messages/bot-text-message');
 const BotTableMessage = require('./messages/bot-table-message');
 const BotfuelAdapter = require('./adapters/botfuel-adapter');
 const BotfuelNlu = require('./nlus/botfuel-nlu');
+const BotfuelTrainerNlu = require('./nlus/botfuel-trainer-nlu');
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/index.js]: This is temporary right?
This should be named botfuel-nlu

File: packages/botfuel-dialog/src/nlus/botfuel-nlu.js [hunk 9]
```diff
@@ -20,11 +20,13 @@ const dir = require('node-dir');
 const Qna = require('botfuel-qna-sdk');
 const logger = require('logtown')('BotfuelNlu');
 const AuthenticationError = require('../errors/authentication-error');
+const ConfigurationError = require('../errors/configuration-error');
 const Classifier = require('../classifier');
 const BooleanExtractor = require('../extractors/boolean-extractor');
 const LocationExtractor = require('../extractors/location-extractor');
 const CompositeExtractor = require('../extractors/composite-extractor');
 const Nlu = require('./nlu');
+const Intent = require('./intent');
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/nlus/botfuel-nlu.js]: Find a different name
Maybe ClassificationResult

File: packages/botfuel-dialog/src/nlus/botfuel-trainer-nlu.js [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * NLU using Botfuel Trainer API
+ */
+class BotfuelTrainerNlu extends Nlu {
+  /** @inheritdoc */
+  constructor(config) {
+    logger.debug('constructor', config);
+    super(config);
+    this.extractor = null;
+    this.qna = null;
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/nlus/botfuel-trainer-nlu.js]: this is useless

File: packages/botfuel-dialog/src/nlus/botfuel-trainer-nlu.js [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+    if (!process.env.BOTFUEL_APP_TOKEN) {
+      throw new SdkError('BOTFUEL_APP_TOKEN are required for using the nlu service');
+    }
+
+    if (!validUrl.isUri(config.trainerApiUri)) {
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/nlus/botfuel-trainer-nlu.js]: Is this temporary?
Config should not define trainerApiUri ...

File: packages/botfuel-dialog/src/nlus/intent.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const logger = require('logtown')('Intent');
+const SdkError = require('../errors/sdk-error');
+
+/** Intent class */
+class Intent {
```
Comment (yangeorget) [inline packages/botfuel-dialog/src/nlus/intent.js]: I am not sure to understand why this class is necessary

File: packages/botfuel-dialog/src/config.js (hunk 4 above)
Comment (tcnguyen) [reply packages/botfuel-dialog/src/config.js]: Yes this is temporary. The purpose of this work is to make things as ready as possible for the trainer official release.

File: packages/botfuel-dialog/src/index.js (hunk 8 above)
Comment (tcnguyen) [reply packages/botfuel-dialog/src/index.js]: Yes this is temporally and will replace botfuel-nlu in the future. If we named in botfuel-nlu now then we must rename current (modified) botfuel-nlu to something else (like natural-nlu) and update this everywhere (default config + tests + may be sample bots that mentioned botfuel-nlu + sample-module)

File: packages/botfuel-dialog/src/nlus/botfuel-trainer-nlu.js (hunk 11 above)
Comment (tcnguyen) [reply packages/botfuel-dialog/src/nlus/botfuel-trainer-nlu.js]: Ok agree. I propose process.env.BOTFUEL_TRAINER_API_URL || "https://...." . What do you think?

**Pull request merged**